src/libsyntax/tokenstream.rs

```diff
@@ -417,112 +462,172 @@ fn covering_span(trees: &[TokenTree]) -> Span {
-                let val = TokLit::Str_(token::intern(&val));
-                Some(TokenStream::from_tts(vec![TokenTree::Token(lit.span,
-                                                                 Token::Literal(val, None))]))
-            }
-            _ => None,
-        }
-
+    /// Copies all of the TokenTrees from the TokenSlice, appending them to the stream.
+    pub fn concat(left: TokenStream, right: TokenStream) -> TokenStream {
+      let new_len = left.len() + right.len();
+      let new_span = combine_spans(left.span(), right.span());
+      TokenStream::Node{
+        left : Rc::new(left),
+        right: Rc::new(right),
+        len: new_len,
+        sp: new_span,
+      }
+    }
+
+    /// Indicate if the TokenStream is empty.
+    pub fn is_empty(&self) -> bool {
+      self.len() == 0
     }
 
-    /// Convert a vector of TokenTrees into a parentheses-delimited TokenStream.
-    pub fn as_paren_delimited_stream(tts: Vec<TokenTree>) -> TokenStream {
-        let new_sp = covering_span(&tts);
-
-        let new_delim = Rc::new(Delimited {
-            delim: token::DelimToken::Paren,
-            open_span: DUMMY_SP,
-            tts: tts,
-            close_span: DUMMY_SP,
-        });
-
-        TokenStream::from_tts(vec![TokenTree::Delimited(new_sp, new_delim)])
+    /// Return a TokenSlice's length.
+    pub fn len(&self) -> usize {
+      match *self {
+        TokenStream::Empty => 0,
+        TokenStream::Leaf { ref tts, .. } => tts.len(),
+        TokenStream::SubLeaf { len, .. } => len,
+        TokenStream::Node { len, .. } => len,
+      }
+    }
+
+    /// Convert a TokenStream into a vector of borrowed TokenTrees.
+    pub fn to_vec(&self) -> Vec<&TokenTree> {
+      match *self {
+        TokenStream::Empty => Vec::new(),
+        TokenStream::Leaf { ref tts, .. } => {
+          let mut v = Vec::new();
+          for t in tts.iter() {
+            v.push(t);
+          }
+          v
+        }
+        TokenStream::SubLeaf { ref tts, offset, len, .. } => {
+          let mut v = Vec::new();
+          for i in offset..(offset+len-1) {
+            v.push(tts.get(i).unwrap())
+          }
+          v
+        }
+        TokenStream::Node { ref left, ref right, .. } => {
+          let mut v1 = TokenStream::to_vec(left);
+          let mut v2 = TokenStream::to_vec(right);
+          v1.append(&mut v2);
+          v1
+        }
+      }
     }
 
-    /// Convert an interned string into a one-element TokenStream.
-    pub fn from_interned_string_as_ident(s: InternedString) -> TokenStream {
-        TokenStream::from_tts(vec![TokenTree::Token(DUMMY_SP,
-                                                    Token::Ident(token::str_to_ident(&s[..])))])
+    /// Convert a TokenStream into a vector of TokenTrees (by cloning the TokenTrees).
+    /// (This operation is expensive.)
+    pub fn to_tts(&self) -> Vec<TokenTree> {
+      TokenStream::to_vec(self).into_iter().map(|x| x.clone()).collect::<Vec<TokenTree>>()
```

Comment:
This can be simplified down into: `TokenStream::to_vec(self).into_iter().cloned().collect()`, but shouldn't you be using the `TokenStream` iterator here instead of `to_vec`?